File: Problems_1-10/Problem3/Problem3.cpp
//Problem 3 :: Largest Prime Factor
#include <iostream>
#include <math.h>

bool is_prime(int num){
    if(num <= 1) return false;
    if(num <= 3) return true;
    if(num % 3 == 0 || num % 2 == 0) return false;
    for (int i=5;i*i<=num;i+=6){
        if (num % i == 0 || num % (i + 2) == 0) return false;
    }
    return true;
}

int main(){
    for(int i=sqrt(600851475143);i>0;i--){
        if((is_prime(i) == true) && (600851475143 % i == 0)){ 
            std::cout << i;
            break;
        }
    }
    return 0;
}

File: Problems_1-10/Problem3/Problem3.java
// Problem 3 :: Largest Prime Factor

public class Main{
    
    private static boolean isPrime(long l){
        if(l < 1) return false;
        if(l <= 3) return true;
        if(l % 2 == 0 || l % 3 == 0) return false;
        for(long i=5;i<(long)Math.sqrt(l);i+=6){
            if(l % i == 0 || l % (i+2) == 0) return false; 
        }
        return true;
    }
    
    public static void main(String args[]){
        long largest_prime = 0;
        for(long i=(long)Math.sqrt(600851475143L);i>0;i--){
            if((isPrime(i) == true) && (600851475143L % i == 0)){ 
                largest_prime = i;
                break;
            }
        }
        System.out.println(largest_prime);
    }
}
